github/auth.go

```diff
@@ -39,183 +37,194 @@ const (
 	patUsername = "git"
 )
 
-var (
-	// ErrInvalidClientOptions is the error returned when calling NewClient() with
-	// invalid options (e.g. specifying mutually exclusive options).
-	ErrInvalidClientOptions = errors.New("invalid options given to NewClient()")
-	// ErrDestructiveCallDisallowed happens when the client isn't set up with WithDestructiveAPICalls()
-	// but a destructive action is called.
-	ErrDestructiveCallDisallowed = errors.New("a destructive call was blocked because it wasn't allowed by the client")
-)
+// ClientOption is the interface to implement for passing options to NewClient.
+// The clientOptions struct is private to force usage of the With... functions.
+type ClientOption interface {
+	// ApplyToGithubClientOptions applies set fields of this object into target.
+	ApplyToGithubClientOptions(target *clientOptions) error
+}
 
-// clientOptions is the struct that tracks data about what options have been set
-// It is private so that the user must use the With... functions.
+// clientOptions is the struct that tracks data about what options have been set.
 type clientOptions struct {
-	// Domain specifies the backing domain, which can be arbitrary if the user uses
-	// GitHub Enterprise. If unset, defaultDomain will be used.
-	Domain *string
-
-	// authTransportFactory is a way to acquire a http.RoundTripper with auth credentials configured.
-	authTransportFactory authTransportFactory
-
-	// RoundTripperFactory is a factory to get a http.RoundTripper that is sitting between the *github.Client's
-	// internal *http.Client, and the *httpcache.Transport RoundTripper. It can be set
-	// for doing arbitrary modifications to http requests.
-	RoundTripperFactory RoundTripperFactory
+	// clientOptions shares all the common options
+	gitprovider.CommonClientOptions
 
-	// EnableDestructiveAPICalls is a flag to tell whether destructive API calls like
-	// deleting a repository and such is allowed. Default: false
-	EnableDestructiveAPICalls *bool
+	// AuthTransport is a ChainableRoundTripperFunc adding authentication credentials to the transport chain.
+	AuthTransport gitprovider.ChainableRoundTripperFunc
 
 	// EnableConditionalRequests will be set if conditional requests should be used.
+	// TODO: Move this to gitprovider.CommonClientOptions if other providers support this too.
```

Comment:
Note for @dinosk in #8. If GitLab supports Conditional Requests, we can make this common.